vaadin-text-field-flow-integration-tests/src/main/java/com/vaadin/flow/component/textfield/tests/BinderValidationPage.java

```diff
@@ -0,0 +1,133 @@
+/*
+ * Copyright 2000-2019 Vaadin Ltd.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not
+ * use this file except in compliance with the License. You may obtain a copy of
+ * the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+package com.vaadin.flow.component.textfield.tests;
+
+import java.math.BigDecimal;
+import java.util.function.Consumer;
+
+import com.vaadin.flow.component.AbstractSinglePropertyField;
+import com.vaadin.flow.component.html.Div;
+import com.vaadin.flow.component.textfield.BigDecimalField;
+import com.vaadin.flow.component.textfield.EmailField;
+import com.vaadin.flow.component.textfield.IntegerField;
+import com.vaadin.flow.component.textfield.NumberField;
+import com.vaadin.flow.component.textfield.PasswordField;
+import com.vaadin.flow.component.textfield.TextArea;
+import com.vaadin.flow.component.textfield.TextField;
+import com.vaadin.flow.data.binder.Binder;
+import com.vaadin.flow.data.binder.Setter;
+import com.vaadin.flow.function.SerializablePredicate;
+import com.vaadin.flow.function.ValueProvider;
+import com.vaadin.flow.router.Route;
+
+@Route("binder-validation")
+public class BinderValidationPage extends Div {
+
+    private Binder<Bean> binder = new Binder<>(Bean.class);
+
+    public static final String BINDER_ERROR_MSG = "binder";
+
+    public BinderValidationPage() {
+
+        addComponent(new TextField(), Bean::getString, Bean::setString,
+                value -> value.length() > 2, field -> field.setMinLength(1));
```

Comment:
If i change this condition to `value.length() == 0 || value.length() > 2` the field doesn't get re-validated once it's invalid and then emptied.

Reply:
I would assume it's because `TextField`'s empty value is `null`.

Reply:
Sorry, no it's not. I'll investigate more.

Reply:
It comes from here: https://github.com/vaadin/vaadin-text-field-flow/blob/3f30431ef704aa5816c83d72dcd5955ae4d1b362/vaadin-text-field-flow/src/main/java/com/vaadin/flow/component/textfield/TextField.java#L56-L61

When the field is empty, the client validation says it's valid.

Looks like we have to consider what to do with client-side validation at this point. We could try to rearrange the listeners so that this runs before the Binder validation. But this listener is still dangerous. If the invalid-property changes at client-side it will revalidate only based on the component's internal validation, since it doesn't know about Binder. Maybe we should not synchronize the invalid-property on `invalid-changed` at all. Also, maybe we should always disable client-side validation for Flow components and handle everything at server-side (like in V8).

Reply:
Great catch btw!

Reply:
One more update: there seems to be a problem with re-validation on blurring.

Steps:
- Focus the field
- Type two characters (i.e. `11`)
- Press Enter (value is applied and `binder` error message is shown)
- Blur the field (it's revalidated and the error-message is hidden)

![valdiation](https://user-images.githubusercontent.com/29372419/68117757-dfbb6e00-ff06-11e9-94e6-df5e7d410746.gif)

Reply:
Oh man. When you hit Enter, it sends the value-changed-event and validates at server-side. When you blur, I guess the web component runs validation also then. But the value has not changed, so there's nothing triggering server-side validation.

Sounds like one more reason to always disable client-side validation. It's too hard to keep the validation logics in sync. And the client doesn't know about Binder validators, so that happens anyway after a roundtrip. Maybe it would be best to always do validation after a roundtrip.

Reply:
Sorry, it actually runs the server-side validation, because of the invalid-changed. But in that case it validates only by the component's logic (minlength=2)